sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/models/IndexBatchException.java

```diff
@@ -0,0 +1,122 @@
+     *
+     * @param originBatch The batch that partially failed indexing.
+     * @param keySelector A lambda that retrieves a key value from a given document of type T.
+     * @param <T> The CLR type that maps to the index schema. Instances of this type can be stored as documents
+     * in the index.
+     * @return A new batch containing all the actions from the given batch that failed and should be retried.
+     */
+    public <T> IndexBatchBase<T> findFailedActionsToRetry(IndexBatchBase<T> originBatch,
+        Function<T, String> keySelector) {
+        List<IndexAction<T>> failedActions = doFindFailedActionsToRetry(originBatch, keySelector);
+        return new IndexBatchBase<T>().setActions(failedActions);
+    }
+
+    /**
+     * @return The indexing results returned by the service.
+     */
+    public List<IndexingResult> getIndexingResults() {
+        return this.results;
+    }
+
+    private static String createMessage(IndexDocumentsResult result) {
+        long failedResultCount = result.getResults().stream()
+            .filter(r -> !r.isSucceeded())
+            .count();
+        return String.format(MESSAGE_FORMAT, failedResultCount, result.getResults().size());
+    }
+
+    private <T> List<IndexAction<T>> doFindFailedActionsToRetry(IndexBatchBase<T> originBatch,
+        Function<T, String> keySelector) {
+        Set<String> uniqueRetriableKeys = getIndexingResults().stream().filter(result ->
+            isRetriableStatusCode(result.getStatusCode())).map(IndexingResult::getKey).collect(Collectors.toSet());
+        return originBatch.getActions().stream().filter(action -> isActionIncluded(action,
+            uniqueRetriableKeys, keySelector))
+            .collect(Collectors.toList());
+    }
+
+    @SuppressWarnings("unchecked")
+    private <T> boolean isActionIncluded(IndexAction<T> action, Set<String> uniqueRetriableKeys,
+        Function<T, String> keySelector) {
+        if (action.getDocument() != null) {
+            return uniqueRetriableKeys.contains(keySelector.apply(action.getDocument()));
+        } else if (action.getParamMap() != null) {
+            return uniqueRetriableKeys.contains(keySelector.apply((T) action.getParamMap()));
+        }
+        return false;
+    }
+
+    private static boolean isRetriableStatusCode(int statusCode) {
+        switch (statusCode) {
+            case 200:
+            case 201:
+                return false;   // Don't retry on success.
+
+            case 404:
+            case 400:
+                return false;   // Don't retry on user error.
+
+            case 500:
+                return false;   // Don't retry when something unexpected happened.
+
+            case 422:
+            case 409:
+            case 503:
+                return true;    // The above cases might succeed on a subsequent retry.
+
+            default:
+                // If this happens, it's a bug. Safest to assume no retry.
+                return false;
+        }
```

Comment:
This could be simplified a lot.

```java
return statusCode == 422 || statusCode == 409 || statusCode == 503;
```

Definitely add a comment for why those status codes are considered for retry but others aren't, for example 503 is allow since the server failed to process the request properly, etc

Reply:
This is copied from .net. Need to double check